apps/state/lib/state/vehicle.ex

```diff
@@ -61,6 +63,38 @@ defmodule State.Vehicle do
     end
   end
 
+  @spec by_labels(labels) :: [Vehicle.t()]
+  def by_labels(labels) do
+    labels
+    |> build_label_matchers()
+    |> select(:label)
+  end
+
+  defp build_label_matchers(labels) do
+    for label <- labels do
+      %{label: label}
+    end
+  end
+
+  @spec by_labels_and_routes(labels, routes, direction_id) :: [Vehicle.t()]
+  def by_labels_and_routes(labels, route_ids, direction_id) do
+    labels
+    |> build_label_and_route_matchers(route_ids, direction_id)
+    |> select(:label)
+  end
+
+  defp build_label_and_route_matchers(labels, route_ids, nil) do
+    for label <- labels, route_id <- route_ids do
+      %{label: label, effective_route_id: route_id}
+    end
+  end
+
+  defp build_label_and_route_matchers(labels, route_ids, direction_id) do
+    for label <- labels, route_id <- route_ids do
+      %{label: label, effective_route_id: route_id, direction_id: direction_id}
+    end
+  end
```

Comment:
You can simplify these different matchers, by using `:_` for fields you aren't using.  Something like:
```elixir
# by_labels is built for you automatically

def by_labels_and_routes(labels, route_ids, direction_id) do
  matchers = build_matchers(route_ids, labels, direction_id || :_)
  select(matchers, :label)
end

def by_route_ids_and_direction_id(route_ids, direction_id) do
  matchers = build_matchers(route_ids, :_, direction_id || :_)
  select(matchers, :route_id)
end

defp build_matchers(route_ids, labels, direction_id) do
  for route_id <- List.wrap(route_ids) label <- List.wrap(labels) do
    %{effective_route_id: route_id, label: label, direction_id: direction_id}
  end
end